M2/Macaulay2/bin/main.cpp

```diff
@@ -0,0 +1,329 @@
+    pushTask(interpTask);
+    waitOnTask(interpTask);
+  }
+  return 0;
+}
+
+/* ######################################################################### */
+
+std::ofstream prof_log;
+thread_local std::vector<char*> M2_stack;
+
+void stack_trace(std::ostream &stream, bool M2) {
+  if(M2) {
+    stream << "M2";
+    for (char* M2_frame : M2_stack)
+      stream << ";" << M2_frame;
+    stream << std::endl;
+  } else {
+    stream << "-* stack trace, pid: " << (long) getpid() << std::endl;
+    stream << boost::stacktrace::stacktrace();
+    stream << "-- end stack trace *-" << std::endl;
+  }
+}
+
+extern "C" {
+  void M2_stack_trace() { stack_trace(std::cout, false); }
+#if PROFILING
+  void M2_stack_push(char* M2_frame) { M2_stack.emplace_back(M2_frame); }
+  void M2_stack_pop() { M2_stack.pop_back(); }
+#else
+  void M2_stack_push(char* M2_frame) {}
+  void M2_stack_pop() {}
+#endif
+}
+
+void* profFunc(ArgCell* p)
+{
+  using namespace std::chrono_literals;
+  std::string filename("profile-" + std::to_string(getpid())+ ".raw");
+  // std::cerr << "Saving profile data in " << filename << std::endl;
+  prof_log.open(filename, std::ios::out | std::ios::trunc );
+  while(true) {
+    std::this_thread::sleep_for(1000ms);
+    tryGlobalTrace();
+  }
+  return NULL;
+}
+
+void* testFunc(ArgCell* p)
+{
+  // TODO: do some math here?
+  printf("testfunc %p\n",p);
+  return NULL;
+}
+
+extern "C" void interp_setupargv(), interp_process();
+extern "C" void reverse_run(struct FUNCTION_CELL *list);
+extern "C" void alarm_handler(int), interrupt_handler(int), trace_handler(int), segv_handler(int);
+
+void* interpFunc(ArgCell* vargs)
+{
+  /* set this as the interpreter's thread? */
+  setInterpThread();
+  /* re-initialize any thread local variables */
+  reverse_run(thread_prepare_list);
+
+  /* setting variables defined in M2.dd */
+  M2_envc = vargs->envc;
+  M2_argc = vargs->argc;
+  M2_envp = M2_tostrings(M2_envc, vargs->envp);
+  M2_argv = M2_tostrings(M2_argc, vargs->argv);
+  M2_args = M2_tostrings(M2_argc == 0 ? 0 : M2_argc - 1, vargs->argv + 1);
+  /* setting commandLine and environment values for frontend */
+  interp_setupargv();
+
+  SETJMP(abort_jmp.addr); /* longjmp to this point when aborted */
+  abort_jmp.is_set = TRUE;
+
+  signal(SIGSEGV, segv_handler);  /* dump the stack trace and exit */
+  signal(SIGUSR1, trace_handler); /* log the stack trace to file */
+
+  /*
+    process() in interp.dd is where all the action happens, however, interp__prepare()
+    from interp-tmp.cc is called first. This happens even before main() because all
+    "_prepare()" functions have "__attribute__ ((constructor))" in their declaration.
+  */
+  interp_process();
+
+  clean_up();
+
+#ifdef MEMDEBUG
+  fprintf(stderr, "gc: heap size = %d, free space divisor = %ld, collections = %ld\n",
+	  GC_get_heap_size(), GC_free_space_divisor, GC_gc_no-old_collections);
+#endif
+
+  exit(0);
+  return NULL;
```

Comment:
This line should not be needed, provided "exit" is given the "noreturn" attribute.  But to avoid
all doubt, replace the pair of lines by "return 0".  Returning NULL is inappropriate, because NULL is for pointers, and main is to return an int.

Reply:
This is from `interpFunc` here:

Reply:
This is simply moved from `interpFunc` here:
https://github.com/Macaulay2/M2/blob/8a473f501318a6d505958ad6332e6618e5f16948/M2/Macaulay2/d/M2lib.c#L609-L612
Is there a reason why `interpFunc` returns a `void*` pointer?

Reply:
Good point.  And I was wrong about it returning an int -- I must have thought the lines were in `main()`.  Gary Furnish must have made it return a pointer so it could be handed to `createThreadTask`.

Leave it as it is for now.  A future experiment would involve removing the "return" and leaving the call to "exit".